tardis/io/logger/logger.py

```diff
@@ -1,165 +1,254 @@
 import logging
-import sys
-
-from tardis.io.logger.colored_logger import ColoredFormatter
-
-logging.captureWarnings(True)
-logger = logging.getLogger("tardis")
-
-console_handler = logging.StreamHandler(sys.stdout)
-console_formatter = ColoredFormatter()
-console_handler.setFormatter(console_formatter)
-
-logger.addHandler(console_handler)
-logging.getLogger("py.warnings").addHandler(console_handler)
-
-LOGGING_LEVELS = {
-    "NOTSET": logging.NOTSET,
-    "DEBUG": logging.DEBUG,
-    "INFO": logging.INFO,
-    "WARNING": logging.WARNING,
-    "ERROR": logging.ERROR,
-    "CRITICAL": logging.CRITICAL,
+import os
+import re
+from dataclasses import dataclass, field
+import panel as pn
+from IPython.display import display
+from functools import lru_cache
+import pandas as pd
+
+PYTHON_WARNINGS_LOGGER = logging.getLogger("py.warnings")
+
+if 'GITHUB_ACTIONS' not in os.environ:
+    pn.extension(comms="ipywidgets")
+else:
+    pn.extension()
+
+@lru_cache(maxsize=1)
+def get_environment():
+    """Determine the execution environment"""
+    if any(x for x in ('VSCODE_PID', 'VSCODE') if x in os.environ):
+        return 'vscode'
+    return 'jupyter'
+
+
+def create_output_widget(height=300):
+    return pn.pane.HTML(
+        "",
+        height=height,
+        styles={
+            'overflow-y': 'auto',
+            'overflow-x': 'auto',
+            'border': '1px solid #ddd',
+            'width': '100%',
+            'font-family': 'monospace',
+            'padding': '8px',
+            'background-color': 'white'
+        }
+    )
+
+ENVIRONMENT = get_environment()
+LOG_OUTPUTS = {
+    "WARNING/ERROR": create_output_widget(),
+    "INFO": create_output_widget(),
+    "DEBUG": create_output_widget(),
+    "ALL": create_output_widget(),
 }
-DEFAULT_LOG_LEVEL = "INFO"
-DEFAULT_SPECIFIC_STATE = False
+TAB_ORDER = ["ALL", "WARNING/ERROR", "INFO", "DEBUG"]
+LOGGER_WIDGET = pn.Tabs(
+    *[(title, LOG_OUTPUTS[title]) for title in TAB_ORDER],
+    height=350,
+    sizing_mode='stretch_width'
+)
+
+@dataclass
+class LoggingConfig:
+    LEVELS: dict = field(default_factory=lambda: {
+        "NOTSET": logging.NOTSET,
+        "DEBUG": logging.DEBUG,
+        "INFO": logging.INFO,
+        "WARNING": logging.WARNING,
+        "ERROR": logging.ERROR,
+        "CRITICAL": logging.CRITICAL,
+    })
+
+    COLORS: dict = field(default_factory=lambda: {
+        logging.INFO: "#D3D3D3",
+        logging.WARNING: "orange",
+        logging.ERROR: "red",
+        logging.CRITICAL: "orange",
+        logging.DEBUG: "blue",
+        "default": "black",
+    })
+
+    DEFAULT_LEVEL = "INFO"
+    DEFAULT_SPECIFIC_STATE = False
+
+LOGGING_LEVELS = LoggingConfig().LEVELS
+
+class AsyncEmitLogHandler(logging.Handler):
```

Comment:
Class needs docstrings

Reply:
In general this file needs updated docstrings

Reply:
Done in https://github.com/tardis-sn/tardis/pull/2872/commits/257f6b16563e1cea84853087696bb95579cdcdc8